client/src/_transport_layer.rs

```diff
@@ -0,0 +1,257 @@
+
+    async fn dispatch(&self, packet: Forwarded) {
+        log::trace!(
+            "[TransportLayer] Dispatching packet from [{}] ({})",
+            packet.node_id,
+            packet.transport
+        );
+
+        match &packet.transport {
+            TransportType::Unreliable => self.dispatch_unreliable(packet).await,
+            TransportType::Reliable => self.virtual_tcp.dispatch(packet).await,
+            // We shouldn't get this transport type from `SessionLayer`, because only TcpLayer
+            // can distinguish packets between `Reliable` and `Transfer`.
+            TransportType::Transfer => self.virtual_tcp.dispatch(packet).await,
+        }
+    }
+
+    pub async fn dispatch_unreliable(&self, forward: Forwarded) {
+        self.ingress_channel.tx.send(forward).ok();
+    }
+
+    async fn spawn_ingress_handler(&self) -> anyhow::Result<()> {
+        let ingress_rx = self
+            .session_layer
+            .receiver()
+            .ok_or_else(|| anyhow::anyhow!("Ingress traffic receiver already spawned"))?;
+
+        tokio::task::spawn_local(self.clone().ingress_handler(ingress_rx));
+        Ok(())
+    }
+
+    async fn ingress_handler(self, ingress_rx: ForwardReceiver) {
+        UnboundedReceiverStream::new(ingress_rx)
+            .for_each(move |forwarded| {
+                let myself = self.clone();
+                async move {
+                    myself.dispatch(forwarded).await;
+                }
+            })
+            .await
+    }
+
+    async fn forward_channel(
+        &self,
+        node_id: NodeId,
+        channel: TransportType,
+    ) -> Option<ForwardSender> {
+        let state = self.state.read().await;
+        match channel {
+            TransportType::Reliable => state.forward_reliable.get(&node_id).cloned(),
+            TransportType::Transfer => state.forward_transfer.get(&node_id).cloned(),
+            TransportType::Unreliable => state.forward_unreliable.get(&node_id).cloned(),
+        }
+    }
+
+    async fn set_forward_channel(
+        &self,
+        node_id: NodeId,
+        channel: TransportType,
+        tx: ForwardSender,
+    ) {
+        let mut state = self.state.write().await;
+        match channel {
+            TransportType::Reliable => state.forward_reliable.insert(node_id, tx),
+            TransportType::Transfer => state.forward_transfer.insert(node_id, tx),
+            TransportType::Unreliable => state.forward_unreliable.insert(node_id, tx),
+        };
+    }
+
+    pub async fn forward(&self, node_id: NodeId) -> anyhow::Result<ForwardSender> {
+        self.forward_generic(node_id, TransportType::Reliable).await
+    }
+
+    pub async fn forward_transfer(&self, node_id: NodeId) -> anyhow::Result<ForwardSender> {
+        self.forward_generic(node_id, TransportType::Transfer).await
+    }
+
+    /// NodeId can be either default or secondary.
+    /// TODO: Make this function resistant to dropping future
+    pub async fn forward_generic(
```

Comment:
```suggestion
    pub async fn forward_through_virtual_tcp(
```

Reply:
Name to discuss, but it's good idea to change

Reply:
`forward_virtual_tcp`